File: MonoScripting/Api/EngineClass/EntityComponents/AudioSourceComponent.cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

public class AudioSourceComponent : IEntityComponent {
    public AudioSourceComponent(ulong handle) : base(handle)
    {
    }

    public void Play(){
        i_Play(handle);
    }
    public void Stop(){
        i_Stop(handle);
    }
    public void Pause(){
        i_Pause(handle);
    }

    public float GetVolume(){
        return i_GetVolume(handle);
    }

    public void SetVolume(float volume){
        i_SetVolume(handle, volume);
    }

    public float GetPitch(){
        return i_GetPitch(handle);
    }

    public void SetPitch(float pitch){
        i_SetPitch(handle, pitch);
    }

    public bool IsLooping(){
        return i_IsLooping(handle);
    }

    public void SetLooping(bool loop){
        i_SetLooping(handle, loop);
    }

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_Play(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_Stop(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_Pause(ulong handle);

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern float i_GetVolume(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_SetVolume(ulong handle, float volume);

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern float i_GetPitch(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_SetPitch(ulong handle, float pitch);

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern bool i_IsLooping(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_SetLooping(ulong handle, bool loop);
}

File: Engine/MonoScripting/Api/EntityComponents/AudioSourceComponent.cpp
#include "../ApiBindings.hpp"
#include <Scene/EntityComponents/AudioSourceComponent.hpp>
#include <MonoScripting/MonoScriptingLayer.hpp>

using namespace VSGE;

static void Play(AudioSourceComponent* ptr) {
	ptr->Play();
}

static void Pause(AudioSourceComponent* ptr) {
	ptr->Pause();
}

static void Stop(AudioSourceComponent* ptr) {
	ptr->Stop();
}

static bool IsPlaying(AudioSourceComponent* ptr) {
	return ptr->IsPlaying();
}

static float GetVolume(AudioSourceComponent* ptr) {
	return ptr->GetVolume();
}
static void SetVolume(AudioSourceComponent* ptr, float volume) {
	ptr->SetVolume(volume);
}

static float GetPitch(AudioSourceComponent* ptr) {
	return ptr->GetVolume();
}
static void SetPitch(AudioSourceComponent* ptr, float pitch) {
	ptr->SetPitch(pitch);
}

static bool IsLooping(AudioSourceComponent* ptr) {
	return ptr->IsLoop();
}
static void SetLooping(AudioSourceComponent* ptr, bool loop) {
	ptr->SetLoop(loop);
}

static void SetAudioClip(AudioSourceComponent* ptr, Resource* res) {
	ptr->SetAudioClip((AudioClipResource*)res);
}

void VSGE::BindAudiosource() {
	MonoScriptingLayer::AddInternalCall("AudioSourceComponent::i_Play(ulong)", Play);
	MonoScriptingLayer::AddInternalCall("AudioSourceComponent::i_Pause(ulong)", Pause);
	MonoScriptingLayer::AddInternalCall("AudioSourceComponent::i_Stop(ulong)", Stop);
	MonoScriptingLayer::AddInternalCall("AudioSourceComponent::i_IsPlaying(ulong)", IsPlaying);

	MonoScriptingLayer::AddInternalCall("AudioSourceComponent::i_GetVolume(ulong)", GetVolume);
	MonoScriptingLayer::AddInternalCall("AudioSourceComponent::i_SetVolume(ulong,float)", SetVolume);

	MonoScriptingLayer::AddInternalCall("AudioSourceComponent::i_GetPitch(ulong)", GetPitch);
	MonoScriptingLayer::AddInternalCall("AudioSourceComponent::i_SetPitch(ulong,float)", SetPitch);

	MonoScriptingLayer::AddInternalCall("AudioSourceComponent::i_IsLooping(ulong)", IsLooping);
	MonoScriptingLayer::AddInternalCall("AudioSourceComponent::i_SetLooping(ulong,bool)", SetLooping);

	MonoScriptingLayer::AddInternalCall("AudioSourceComponent::i_SetAudioClip(ulong,ulong)", SetAudioClip);
}